File: ScriptKit.CLR/NET/Collections/Generic/List.cs
namespace System.Collections.Generic 
{
    [ScriptKit.CLR.Ignore]
    [ScriptKit.CLR.Name("ScriptKit.List")]
    public sealed class List<T> : ICollection<T> 
    {
        public List() 
        {
        }

        public List(IEnumerable<T> items) 
        {
        }

        public int Count 
        {
            get 
            {
                return 0;
            }
        }

        public T Get(int index)
        {
            return default(T);
        }

        public void Set(int index, T value)
        {
        }

        public void Add(T item) 
        {
        }

        public void AddRange(IEnumerable<T> items) 
        {
        }

        public void Clear() 
        {
        }
        
        public bool Contains(T item) 
        {
            return false;
        }

        public IEnumerator<T> GetEnumerator() 
        {
            return null;
        }

        public List<T> GetRange(int index) 
        {
            return null;
        }

        public List<T> GetRange(int index, int count) {
            return null;
        }

        public int IndexOf(T item) 
        {
            return 0;
        }

        public int IndexOf(T item, int startIndex) 
        {
            return 0;
        }

        public void Insert(int index, T item) 
        {
        }

        public void InsertRange(int index, IEnumerable<T> items) 
        {
        }

        public string Join() 
        {
            return null;
        }

        public string Join(string delimiter) 
        {
            return null;
        }

        public int LastIndexOf(object item) 
        {
            return 0;
        }

        public int LastIndexOf(object item, int fromIndex) 
        {
            return 0;
        }

        public bool Remove(T item) 
        {
            return false;
        }

        public void RemoveAt(int index) 
        {
        }

        public List<T> RemoveRange(int index, int count) 
        {
            return null;
        }

        public void Reverse() 
        {
        }

        public List<T> Slice(int start) 
        {
            return null;
        }

        public List<T> Slice(int start, int end) 
        {
            return null;
        }

       
        public void Sort() {
        }

        public void Sort(Func<T, T, int> comparison)
        {
        }

        public void Splice(int start, int deleteCount) 
        {
        }

        public void Splice(int start, int deleteCount, IEnumerable<T> itemsToInsert) 
        {
        }

        public void Unshift(params T[] items) 
        {
        }

        public T[] ToArray()
        {
            return null;
        }
    }
}

File: Core/CLR/resources/collections/List.js
Bridge.Class.generic('Bridge.List$1', function (T) {
    var $$name = Bridge.Class.genericName('Bridge.List$1', T);
    return Bridge.Class.cache[$$name] || (Bridge.Class.cache[$$name] = Bridge.Class.extend($$name, {
        $extend: [Bridge.ICollection$1(T)],
        $init: function (obj) {
            if (Object.prototype.toString.call(obj) === '[object Array]') {
                this.items = obj;
            }
            else if (Bridge.is(obj, Bridge.IEnumerable)) {
                this.items = Bridge.toArray(obj);
            }
            else {
                this.items = [];
            }
        },

        checkIndex: function (index) {
            if (index < 0 || index > (this.items.length - 1)) {
                throw new Error('Index out of range');
            }
        },

        getCount: function () {
            return this.items.length;
        },

        get: function (index) {
            this.checkIndex(index);

            return this.items[index];
        },

        set: function (index, value) {
            this.checkIndex(index);
            this.items[index] = value;
        },

        add: function (value) {
            this.items.push(value);
        },

        addRange: function (items) {
            var array = Bridge.toArray(items),
                i,
                len;

            for (i = 0, len = array.length; i < len; ++i) {
                this.items.push(array[i]);
            }
        },

        clear: function () {
            this.items = [];
        },

        indexOf: function (item, startIndex) {
            var i;

            if (!Bridge.isDefined(startIndex)) {
                startIndex = 0;
            }

            if (startIndex != 0) {
                this.checkIndex(index);
            }

            for (i = startIndex; i < this.items.length; i++) {
                if (item === this.items[i]) {
                    return i;
                }
            }

            return -1;
        },

        contains: function (item) {
            return this.indexOf(item) > -1;
        },

        getEnumerator: function () {
            return new Bridge.ArrayEnumerator(this.items);
        },

        getRange: function (index, count) {
            if (!Bridge.isDefined(index)) {
                index = 0;
            }

            if (!Bridge.isDefined(count)) {
                count = this.items.length;
            }

            if (index != 0) {
                this.checkIndex(index);
            }

            this.checkIndex(index + count - 1);

            var result = [],
                i;

            for (i = index; i < count; i++) {
                result.push(this.items[i]);
            }

            return result;
        },

        insert: function (index, item) {
            if (index != 0) {
                this.checkIndex(index);
            }


            if (Bridge.isArray(item)) {
                for (var i = 0; i < item.length; i++) {
                    this.insert(index++, item[i]);
                }
            }
            else {
                this.items.splice(index, 0, item);
            }
        },

        join: function (delimeter) {
            return this.items.join(delimeter);
        },

        lastIndexOf: function (item, fromIndex) {
            if (!Bridge.isDefined(fromIndex)) {
                fromIndex = this.items.length - 1;
            }

            if (fromIndex != 0) {
                this.checkIndex(fromIndex);
            }

            for (var i = fromIndex; i >= 0; i--) {
                if (item === this.items[i]) {
                    return i;
                }
            }

            return -1;
        },

        remove: function (item) {
            var index = this.indexOf(item);

            this.checkIndex(index);
            this.items.splice(index, 1);
        },

        removeAt: function (index) {
            this.checkIndex(index);
            this.items.splice(index, 1);
        },

        removeRange: function (index, count) {
            this.checkIndex(index);
            this.items.splice(index, count);
        },

        reverse: function () {
            this.items.reverse();
        },

        slice: function (start, end) {
            this.items.slice(start, end);
        },

        sort: function (comparison) {
            this.items.sort(comparison);
        },

        splice: function (start, count, items) {
            this.items.splice(start, count, items);
        },

        unshift: function () {
            this.items.unshift();
        },

        toArray: function () {
            return Bridge.toArray(this);
        }
    }));
});
